Add server selector component

twoplustwoone opened the pull request:
Add the `api-server-selector` component to the top of the `api-request-panel`.

If the `baseUri` property is set on this component, do not show the selector and continue using the set base URI.

twoplustwoone committed:
feat: add server selector and listen for change event

twoplustwoone committed:
Merge branch 'stage' into feat/add-server-selector

jarrodek commented on package.json:
```diff
@@ -30,6 +30,7 @@
     "@advanced-rest-client/headers-parser-mixin": "^3.0.0",
     "@advanced-rest-client/response-view": "^3.0.3",
     "@api-components/api-request-editor": "^4.0.0",
+    "@api-components/api-server-selector": "0.0.1",
```
If we already putting the selector into other component let's make it at least `0.1.0`. Otherwise with every change in the selector you will have to update all the components that are using this selector.

jarrodek commented on src/ApiRequestPanel.js:
```diff
@@ -166,6 +168,14 @@ export class ApiRequestPanel extends EventsTargetMixin(HeadersParserMixin(LitEle
     `;
   }
 
+  _renderServerSelector() {
```
So, what will happen when you select "custom" option in the selector and set an URL? It should update the `baseUri` on the components so they can handle this in the UI. But now when you do this the selector will disappear. This should not work like this. We need to know whether the base URI was an effect of setting it up in the selector or not. 
I guess that the right approach would be to add this information to the event dispatched by the selector and set some flag here. If the base URI is a result of setting it in the selector then the selector stays rendered. Otherwise it is removed. Also, the flag has to be cleared when you actually change the baseUri from the outside. It can't happen automatically but you can dispatch an event on the console to clear this state.

twoplustwoone replied:
This actually doesn't happen the way you're saying. When you select custom and set a URI, the `baseUri` property does not change, the `_selectedServer` property changes. This template function checks if the `baseUri` property is set (which is only being set by the parent), so the server selection change won't modify this value (i.e., I'm never reassigning `this.baseUri`)

jarrodek commented on src/ApiRequestPanel.js:
```diff
@@ -419,6 +430,19 @@ export class ApiRequestPanel extends EventsTargetMixin(HeadersParserMixin(LitEle
+    if (old === value) {
+      return;
+    }
+    this._selectedServer = value;
```
You need to add `this.requestUpdate('selectedServer', value);` otherwise the element won't know that the UI should be re-rendered.

twoplustwoone replied:
`_selectedServer` is declared in the properties, so when we set `selectedServer`, it changes `_selectedServer` and re-renders

jarrodek commented on src/ApiRequestPanel.js:
```diff
@@ -517,6 +543,9 @@ export class ApiRequestPanel extends EventsTargetMixin(HeadersParserMixin(LitEle
     this._appendProxy(e);
   }
+  _serverChangeHandler(e) {
+    this.selectedServer = e.detail.value;
```
So here you would set a variable that sets an information that the base URI is set from the selector and the selector should continue to be rendered. Then you need to clear it when the selection change.

twoplustwoone replied:
Not needed, see my comment above

jarrodek commented on test/api-request-panel.test.js:
```diff
@@ -2,7 +2,7 @@ import { fixture, assert } from '@open-wc/testing';
 import '../api-request-panel.js';
 
-describe('<api-request-panel>', function() {
+describe('<api-request-panel>', function () {
```
We need to agree on common linting rules :)

twoplustwoone replied:
Agreed. I'll undo this

jarrodek commented on test/api-request-panel.test.js:
```diff
@@ -75,6 +79,41 @@ describe('<api-request-panel>', function() {
+        element.selectedServer = 'https://www.google.com';
+        assert.equal(element.selectedServer, 'https://www.example.org');
+      });
+    })
```
This missing a test of what happens when you select a "custom" option in the selector and provide an URI.

twoplustwoone replied:
`api-request-panel` shouldn't care what option is selected, it should need to receive the base URI and apply it to its children. This removes the need to keep track of _which_ server is selected, and only focus on telling its children what the value is

twoplustwoone replied:
But I updated my tests to show this

jarrodek commented on src/ApiRequestPanel.js:
```diff
@@ -130,13 +131,14 @@ export class ApiRequestPanel extends EventsTargetMixin(HeadersParserMixin(LitEle
       .amf="${amf}"
       ?noUrlEditor="${noUrlEditor}"
-      .baseUri="${baseUri}"
+      .baseUri="${selectedServer}"
```
This is not enough. Actually you shouldn't be doing it this way. As we agreed today, you need to compute the model for the server and pass it to the editor. It already handles [server property](https://github.com/advanced-rest-client/api-request-editor/blob/stage/src/ApiRequestEditor.js#L94) made for the partial model. The base URI stays as it was.

jarrodek reviewed (changes requested):
Sorry, but more changes needed.

twoplustwoone committed:
test: add tests

Update api-server-selector to 0.1.0

twoplustwoone committed:
Merge branch 'feat/add-server-selector' of github.com:advanced-rest-client/api-request-panel into feat/add-server-selector

twoplustwoone committed:
feat: add selectedServer and selectedServerType

twoplustwoone committed:
Merge branch 'stage' into feat/add-server-selector

twoplustwoone committed:
chore: update package-lock

jarrodek commented:
So, how the `server` property is now computed? We decided to do server computations based on the selection in `api-documentation` and `api-request-panel`. You are handling selection change from the selector but it does nothing except for setting back the properties to the selector (which internally has the same properties already set). The idea here was to compute the server from current selection and pass it to the editor. Or the base URI if the selection is custom / slotted.

twoplustwoone committed:
feat: add AmfHelperMixin and compute servers

Compute server selection
Add tests
Add AmfHelperMixin

twoplustwoone committed:
test: add amf-helper and amf-loader

twoplustwoone committed:
fix: multiple fixes

Document properties
Remove set and get for amf
Add __amfChanged function
requestUpdate for selectedServerValue
remove raml-aware

twoplustwoone committed:
fix: remove amf from properties, remvoe apiChanged method

jarrodek reviewed (dismissed):
LGTM

twoplustwoone committed:
build: bump version

lbauret reviewed (approved):
LGTM!
